dali/ft/filecopy.cpp

```diff
@@ -3249,13 +3256,117 @@ void FileSprayer::updateTargetProperties()
                 if (stricmp(aname, "Protect") != 0)
                     curProps.addPropTree(aname, createPTreeFromIPT(&iter->query()));
             }
+
+            // Handle history
+            IPropertyTree *curHistory = lock.queryHistory();
+
+            if (srcHistory)
+            {
+                // copy history records from source file
+                Owned<IPropertyTreeIterator> historyIter = srcHistory->getElements("*");
+                ForEach(*historyIter)
+                {
+                    const char *aname = historyIter->query().queryName();
+                    curHistory->addPropTree(aname, createPTreeFromIPT(&historyIter->query()));
+                }
+            }
+
+            // Add new record about this operation
+            IPropertyTree * newRecord = createPTree("Origin");
+
+            CDateTime temp;
+            temp.setNow();
+            unsigned hour, min, sec, nanosec;
+            temp.getTime(hour, min, sec, nanosec);
+            temp.setTime(hour, min, sec, 0);
+
+            StringBuffer timestr;
+            newRecord->addProp("@timestamp",temp.getString(timestr).str());
+
+            StringBuffer val;
+            srcAttr->getProp("@owner", val);
+            newRecord->addProp("@owner", val.str());
+
+            srcAttr->getProp("@workunit", val.clear());
+            if (0 == val.length())
+                val.set("N/A");
+            newRecord->addProp("@workunit", val.str());
+
+            newRecord->addProp("@operation", getOperationTypeString());
+
+            // add original file name
+            if (distributedSource)
+            {
+                RemoteFilename remoteFile;
+                distributedSource->getPart(0)->getFilename(remoteFile, 0 );
+                splitAndStoreFileInfo(newRecord, remoteFile);
+            }
+            else
+            {
+                val.clear();
+                ForEachItemIn(idx, sources)
+                {
+                    FilePartInfo & curSource = sources.item(idx);
+                    RemoteFilename remoteFile = curSource.filename;
+                    splitAndStoreFileInfo(newRecord, remoteFile, idx, false);
```

Comment:
I think the mixed used of splitAndStoreFileInfo for single and multi source handling is a bit confusing.
Handling the 1st part (0 == idx) to handle the drive/path/ip extraction vs others... followed by different code for each case.

It would be clearer if you did that 1st unconditionally (before line 3298) as both these branch have to do.
Then dealt with each case separately, as code is not the same.